### PR TITLE
[WIP] Introduce crystal --fake_source=sourcepath flag

### LINKED ISSUE #4561: Compile source from STDIN
---
**Suggested tags:** `compiler` `feature`

---

Given how [vim-crystal](https://github.com/rhysd/vim-crystal) and [ale](https://github.com/w0rp/ale) are built for vim, there is a syntax checker for crystal that is built by using the compiler. The problem I've faced with both of these plugins for vim/neovim are that the compiler cannot run against a file that does not exist.

I've begun hacking on _ale_ in order to improve the Crystal support for it, but it appears to be an issue that the compiler itself could help with. I'd like to:

```
$ echo 'puts "hello world!"' | crystal build --no-codegen --no-color -o /dev/null -f json -
```

The problem is that it tries to interpret the file named `-` as an actual file instead of reading the file contents from STDIN.

I understand this is a unix convention and we're focusing a lot more on cross-platform compatibility with even Windows support, but I feel like solidifying editor support would be a good way of doing that. Thoughts?

### PR DESCRIPTION
**Pull request by akzhan**
to compile source from STDIN, closes #4561.

## Synopsys

```bash
cat src/math/math.cr | ./bin/crystal build --no-codegen --no-color -o /dev/null -f json --fake-source src/math/math.cr
```

### CODE REVIEW

File: src/compiler/crystal/command.cr [hunk 1]
```diff
@@ -450,6 +450,9 @@ class Crystal::Command
   end
 
   private def gather_sources(filenames)
+    if ENV.has_key?("CRYSTAL_FAKE_SOURCE_FILENAME")
```
Comment (Sija) [inline src/compiler/crystal/command.cr]: I'd suggest getting rid of extra `ENV#[]` call, like:

```cr
if fake_source_filename = ENV["CRYSTAL_FAKE_SOURCE_FILENAME"]?
  return [Compiler::Source.new(fake_source_filename, STDIN.gets_to_end)]
end
```

**Comment by RX14** [comment]
This seems like a weird solution. I don't mind `CRYSTAL_FAKE_SOURCE_FILENAME` but it shouldn't be required just to compile a  file from stdin. The default should just be `stdin` as the filename with requires relative to the cwd.

**Comment by akzhan** [comment]
@RX14 Yes, it is just a proposal (how to implement). Now it is `--fake-source` compiler flag.

File: src/compiler/crystal/command.cr (hunk 1 above)
Comment (akzhan) [inline src/compiler/crystal/command.cr]: just reworked as `--fake-source` flag. looks cleaner.

**Comment by akzhan** [comment]
Followed with #4571 with cleaner code.

**Pull request closed without merging**